src/main/scala/org/broadinstitute/dsde/workbench/sam/model/Notification.scala

```diff
@@ -0,0 +1,80 @@
+
+  sealed abstract class NotificationType[T <: Notification: TypeTag] {
+    def baseKey = Notifications.baseKey[T]
+    val format: RootJsonFormat[T]
+    val notificationType = typeOf[T].typeSymbol.asClass.name.toString
+    val description: String
+
+    /** means the user can never turn it off */
+    val alwaysOn = false
+  }
+
+  sealed trait Notification {
+    def key = Notifications.baseKey(this)
+  }
+
+  private val allNotificationTypesBuilder = Map.newBuilder[String, NotificationType[_ <: Notification]]
+
+  /**
+   * called internally to register a notification type so it will appear in the allNotificationTypes map
+   * @param notificationType
+   * @tparam T
+   * @return notificationType
+   */
+  private def register[T <: Notification](notificationType: NotificationType[T]): NotificationType[T] = {
+    require(allNotificationTypes == null, "all calls to register must come before definition of allNotificationTypes in the file")
+    allNotificationTypesBuilder += notificationType.notificationType -> notificationType
+    notificationType
+  }
+
+  case class GroupAccessRequestNotification(recipientUserId: WorkbenchUserId, groupName: WorkbenchGroupName, replyToIds: Set[WorkbenchUserId], requesterId: WorkbenchUserId) extends Notification
+  val GroupAccessRequestNotificationType = register(new NotificationType[GroupAccessRequestNotification] {
+    override val format = jsonFormat4(GroupAccessRequestNotification)
+    override val description = "Group Access Requested"
+  })
+
+  // IMPORTANT that this comes after all the calls to register
+  val allNotificationTypes: Map[String, NotificationType[_ <: Notification]] = allNotificationTypesBuilder.result()
+
+  implicit object NotificationFormat extends RootJsonFormat[Notification] {
+
+    private val notificationTypeAttribute = "notificationType"
+
+    override def write(obj: Notification): JsValue = {
+      val notificationType = obj.getClass.getSimpleName
+      val json = obj.toJson(allNotificationTypes.getOrElse(notificationType, throw new SerializationException(s"format missing for $obj")).format.asInstanceOf[RootJsonWriter[Notification]])
+
+      JsObject(json.asJsObject.fields + (notificationTypeAttribute -> JsString(notificationType)))
+    }
+
+    override def read(json: JsValue) : Notification = json match {
+      case JsObject(fields) =>
+        val notificationType = fields.getOrElse(notificationTypeAttribute, throw new DeserializationException(s"missing $notificationTypeAttribute property"))
+        notificationType match {
+          case JsString(tpe) => allNotificationTypes.getOrElse(tpe, throw new DeserializationException(s"unrecognized notification type: $tpe")).format.read(json)
+          case x => throw new DeserializationException(s"unrecognized $notificationTypeAttribute: $x")
+        }
+
+      case _ => throw new DeserializationException("unexpected json type")
+    }
+  }
+}
```

Comment:
If you think you should move `NotificationDao.scala` to `workbench-libs`, should this move with it?  I know `GroupAccessRequestNotification` stuff seems SAM specific, but could this be externalized in some say so that the rest of this generic code could be extracted?

Reply:
In general, this feels like a lot generalizable code being added for what seems like limited/simple/basic functionality.    What does SAM really need?  Do we foresee SAM needing to send other notifications?  

In general, I feel like there's something here that I'm having a hard time articulating.  I guess it's that there's a lot of logic that you're adding here to support general Notifications via pub/sub and we're only doing a specific Notification in SAM.  So perhaps my suggestion is to either go ahead and do the extraction to `workbench-libs` or maybe simplify the logic in SAM so that it's specific to the one Notification we're trying to send here.